### PR TITLE
Make `StridedReinterpretArray`'s `get/setindex` pointer based.

### PR DESCRIPTION
**Pull request by N5N3**
This PR make `StridedReinterpretArray`'s `get/setindex` pure pointer based if its root parent is a `Array`.
Thus a "Dense" `ReinterpretArray` should behave more like a `Array`.
Some examples with better performance:
```julia
julia> a = randn(ComplexF64, 100, 100); b = randn(100); c = a * b;
julia> aa = reinterpret(Float64, a); cc = reinterpret(Float64, c);
julia> @btime LinearAlgebra.generic_matvecmul!($cc, 'N', $aa, $b, LinearAlgebra.MulAddMul(true,false));
  2.544 μs (0 allocations: 0 bytes)  # on master 116.900 μs (0 allocations: 0 bytes)
```
```julia
julia> f(x, y) = @inbounds @simd ivdep for i in eachindex(x,y) # ivdep here is useless on master
       x[i] = ntoh(y[i])
       end

julia> a = reinterpret(Float64, rand(UInt8, 16000));

julia> @btime f($a, $a)
  124.388 ns (0 allocations: 0 bytes) # on master 697.260 ns (0 allocations: 0 bytes)
```
Test has been extended thus all branches should be tested.

### CODE REVIEW

**Comment by jishnub** [comment]
@N5N3 this PR seems to provide a significant performance boost for reinterpreted arrays. If you're happy with this, would you mind recommending some reviewers so that it might move ahead?

**Comment by jishnub** [comment]
Gentle bump

**Comment by ronisbr** [comment]
One possible use case that might benefit from this PR:

https://discourse.julialang.org/t/slowdown-with-reinterpret/91749

**Comment by j-fu** [comment]
In fact reinterpret helps a lot when different approaches to handle e.g. coordinate arrays for discretization grids need to be reconciled. E.g. mesh generators written in C/C++ may return a 3xn matrix of coordinates, and this could be readily reinterpreted e.g. as a Vector{Point3}, and vice versa. So I think the use case is not very exotic.

**Comment by gbaraldi** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/44186#issuecomment-1363278008) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2cb1439_vs_b6f32bc/report.html).

**Comment by Moelf** [comment]
bump, this helps greatly for some of the long-standing performance issues w.r.t. `StridedReinterpretArray`

**Comment by vtjnash** [comment]
I removed the `unsafe_convert` method added here, since it does not seem correct, vis a vis 0a0bd007ab1. Let's see how CI is doing

File: test/testhelpers/arrayindexingtypes.jl [hunk 1]
```diff
@@ -66,3 +66,6 @@ Base.axes(A::WrapperArray) = axes(A.parent)
 Base.getindex(A::WrapperArray, i::Int...) = A.parent[i...]
 Base.setindex!(A::WrapperArray, v, i::Int...) = A.parent[i...] = v
 Base.similar(A::WrapperArray, ::Type{T}, dims::Dims) where T = similar(A.parent, T, dims)
+Base.unsafe_convert(::Type{Ptr{T}}, A::WrapperArray{T}) where {T} = Base.unsafe_convert(Ptr{T}, A.parent)
```
Comment (N5N3) [inline test/testhelpers/arrayindexingtypes.jl:69]:
```suggestion
Base.cconvert(::Type{Ptr{T}}, A::WrapperArray{T}) where {T} = Base.cconvert(Ptr{T}, A.parent)
```

File: test/reinterpretarray.jl [hunk 2]
```diff
@@ -534,4 +587,4 @@ end
     @test reinterpret(Bytes15, (Int8(1), Int16(2), Int32(3), Int64(4))) == Bytes15(Int8(1), Int16(2), Int32(3), Int64(4))
 
     @test_throws ArgumentError reinterpret(Tuple{Int32, Int64}, (Int16(1), Int64(4)))
-end
+end
```
Comment (giordano) [inline test/reinterpretarray.jl:590]:
```suggestion
end

```

**Comment by N5N3** [comment]
Unfortunately, I can confirm this PR blocks Vectorization in some simple cases.

**Comment by vtjnash** [comment]
Ah okay. Is that because this intercepts at sort of a bad point, so rather than doing all-native load/store/bitcast for simple cases like Int64<->Float64, now those always go through our unsafe primitives, that are only supposed to be for C-compatibility, of unsafe load/store, and then lose out on the efficiency benefits of the native primitives?

**Comment by N5N3** [comment]
I thought LLVM might prefer `unsafe_store!(p, v, li)` rather than `unsafe_store!(p + sizeof(T) * (li - 1), v)`.
But the "regression" is not fixed.
`unsafe_store!(p, v, li)` is fully intrinsiced based, and it does make LLVM IR simpler, but LLVM still refused to generate Vectorization block.

**Comment by vtjnash** [comment]
I think there is roughly 2 cases here:
 - isprimitivetype for both source and dest
 - only isbitstype

For isprimitivetype, it looks like we might be only using native operations currently, and those will end in LLVM as a load/store into registers. That seems the case where it looks like it should be already capable of emitting the optimal code, so this can only make it worse by using unsafe operations which requires the compiler to destroy useful information (e.g. vectorizability).

For everything else, it looks like we always use a native load from the array too, but for other cases we know that will turn into a memcpy to a stack slot, then we do a move from that stack slot to a Ref, then we do an unsafe_load of that Ref, which we know will also turn into a memcpy to yet another stack slot, and finally we return that, which should do one final copy from that stack slot to the sret. After inlining and optimizations, most of those copies should go away (those are the same copies we would expect if this wasn't a reinterpret, but merely a copy via a `Ref(x)[]`, so this should not be too unusual of IR)

**Comment by vtjnash** [comment]
Sure, but we know doing stuff with `unsafe_` operations is expected to perform badly. The advantage of this PR seems to be that since it was already using the bad `unsafe_` operations in many cases, that sometimes there are ways to make those more direct (with `unsafe_load`) instead of waiting for LLVM to transform them to something roughly equivalent

**Comment by vtjnash** [comment]
We should probably now define that `parent(::Array) = getfield(array.ref.mem)`, like all the other types. But a viewed/reshaped Memory typically _is_ an Array. Although maybe you meant you also needed it to be a non-mutable object?

**Comment by N5N3** [comment]
>Although maybe you meant you also needed it to be a non-mutable object?

Yes, I mean `Base.SubArray` and `Base.ReshapedArray`

File: base/reinterpretarray.jl [hunk 3]
```diff
@@ -352,23 +352,32 @@ has_offset_axes(a::ReinterpretArray) = has_offset_axes(a.parent)
...
-@inline @propagate_inbounds getindex(a::ReinterpretArray) = a[firstindex(a)]
+check_ptr_indexable(a::ReinterpretArray, sz = elsize(a)) = check_ptr_indexable(parent(a), sz)
+check_ptr_indexable(a::ReshapedArray, sz) = check_ptr_indexable(parent(a), sz)
+check_ptr_indexable(a::FastContiguousSubArray, sz) = check_ptr_indexable(parent(a), sz)
+check_ptr_indexable(a::Array, sz) = sizeof(eltype(a)) !== sz
+check_ptr_indexable(a::Memory, sz) = true
+check_ptr_indexable(a::AbstractArray, sz) = false
 
-@inline @propagate_inbounds function getindex(a::ReinterpretArray{T,N,S}, inds::Vararg{Int, N}) where {T,N,S}
+@propagate_inbounds getindex(a::ReinterpretArray) = a[firstindex(a)]
```
Comment (Moelf) [inline base/reinterpretarray.jl:370]: in retrospect this looks like a bug #58232

**Pull request merged**